sympy/ntheory/factor_.py

```diff
@@ -836,7 +989,7 @@ def _check_termination(factors, n, limitp1, use_trial, use_rho, use_pm1,
         for b, e in facs.items():
             if verbose:
                 print(factor_msg % (b, e))
-            factors[b] = exp*e
+            factors[b] = int(exp*e)
```

Comment:
```suggestion
            factors[b] = exp*e
```

Reply:
I had to convert it to an int because it would have resulted in an error.
Upon closer examination of this error, flint was the cause.

flintlib/python-flint#92

How should we handle this on the sympy side?

Reply:
Add a test that uses the flint func that gives an error -- nine int -- and assert that the result should be int. Xfail it. Put a note in it that says "when this passes remove int wrapper on return value in _perfect_power". Or on all the imported flint versions of functions, wrap their output in int...`gcd=lambda *x: int(_flintgcd(*x))`

Reply:
Looking at the error report, though, I don't see the connection since flint would return 1 in some cases instead of expected answer. How does using int solve that?

Reply:
I think it would be better to perform the conversion in the place where 3-arg pow is being used so it is clear that that is where the potential problem is.

Reply:
This is where the problem occurs.

https://github.com/sympy/sympy/blob/6e347f05210519d5eef3f8b894ccf5c4d7029226/sympy/ntheory/residue_ntheory.py#L1830

If this is changed to `pow(p, e0, int(modulo))`, the error will no longer occur.

Reply:
A separate question is why is it that the types are mixed i.e. should it not be either `pow(int, int, int)` or `pow(fmpz, fmpz, fmpz)`. If we have `pow(int, int, fmpz)` then does that not mean that a conversion is missing somewhere?

Reply:
Since modulo was calculated from pow [here](https://github.com/sympy/sympy/blob/6e347f05210519d5eef3f8b894ccf5c4d7029226/sympy/ntheory/residue_ntheory.py#L1745) -- and is apparently not an int? -- the `pow` is faulty. But `pow` is a built in, so how is this problem being created?

Reply:
In `_perfect_power`, `e` in `factors[b] = exp*e` may be `fmpz` because of `sympy.external.gmpy.gcd`. Therefore, the value of `factorint()` is `fmpz`.

Reply:
Then is the solution to write the following in external?
```python
from sympy.external.gmpy import gcd as _gcd
def gcd(a,b):  # or whatever the signature is
    return int(_gcd(a,b))  # fix for issue ...
```

Reply:
If we look at the part in issue in this case, we can see that it is called roughly in the following hierarchy.

* `binomial_mod`
* `factorint`
* `_perfect_power`
* `sympy.external.gmpy.gcd`

The issue now is at what layer to convert to int. If we have a consistent policy, we can convert anywhere, but if we do it at too many lower layers, I am concerned about the overhead of the conversion. I think it would be better if the return value of public functions were unified to int. If we follow that policy, I think we should just set the return value of `factorint` to int.

Reply:
There is nothing wrong with `gmpy` or `gmpy.gcd`. Also, I did not understand the relation between the bug being fixed and the overhead being eliminated.

Reply:
Sorry. Ignore everything I said and let me start over.

[Here](https://github.com/sympy/sympy/pull/25749#discussion_r1342545334) you said it was a flint error that happens when the args are not all of the same type. If that's the case then we need to make sure that we never pass mixed args to `pow`. When will that happen? How are we going to track that? @oscarbenjamin has thought a lot about how to work with matrices in different domains. I have a hard time believing that in all of ntheory that factorint is the only function that might output a mixture of int and FPZ (or whatever the offending type is).

Reply:
This bug will be resolved in the next version of flint. Permanently, I think it would be better to restrict flint to be used only in later versions. In the short term, we can avoid the error by setting the return value of factorint to int.

> I have a hard time believing that in all of ntheory that factorint is the only function that might output a mixture of int and FPZ (or whatever the offending type is).

I don't think this is taken into account in many of ntheory's functions, but unless there is a serious bug like this one, I don't think it will be fatal. I guess it's just a matter of fixing what we notice as we go along.

Reply:
We do need to have a clearer idea of which interfaces are supposed to accept and return int and which are supposed to accept and return other types like mpz or fmpz. The mpz type from gmpy2 is designed to be interchangeable with int for the most part (there are some differences) but fmpz is not e.g.:
```python
In [2]: flint.fmpz(2) / flint.fmpz(3)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[2], line 1
----> 1 flint.fmpz(2) / flint.fmpz(3)

TypeError: unsupported operand type(s) for /: 'flint.types.fmpz.fmpz' and 'flint.types.fmpz.fmpz'
```
Ideally all internal code would operate with a clear choice of one type. This happens cleanly in the poly domains because we have explicit conversions into and out of each domain. In the case of ntheory it is all blurry which functions are internal and which are for end users which makes this more difficult.

Perhaps it would be cleaner if internal ntheory functions used the ZZ domain explicitly. The ZZ domain is already designed to provide a transparent interface over the different ground types.

Reply:
For now, then, the problem with fixing it [here](https://github.com/sympy/sympy/pull/25749/files#r1342136241) is that there is no bell to go off when the resolution has been fixed. So that is why I recommend a "bell test". If you write a test in the test suite that asserts the current state with a note that indicates that if it fails then the `int()` wrapper in `factorint` should be removed, then this won't get forgotten:
```python
def test_flint_int_issue():
    # when this fails, the int wrapper is no longer needed in factorint
    assert {condition that exists that makes wrapping in int necessary}
```